gui/include/arbitration_graphs/gui/web_server.hpp

```diff
@@ -77,11 +87,42 @@ class WebServer {
     }
 
 private:
+    std::string dataDirectory() {
+        namespace fs = std::filesystem;
+
+        // Check if the install directory works
+        if (fs::exists(INSTALL_APP_DIRECTORY) && fs::is_directory(INSTALL_APP_DIRECTORY)) {
+            return INSTALL_APP_DIRECTORY;
+        }
+
+        // Check if the local source directory works
+        if (fs::exists(LOCAL_APP_DIRECTORY) && fs::is_directory(LOCAL_APP_DIRECTORY)) {
+            return LOCAL_APP_DIRECTORY;
+        }
+
+        // Check the APP_DIRECTORY environment variable
+        if (const char* pathEnv = std::getenv("APP_DIRECTORY")) {
+            fs::path dir = fs::path(pathEnv);
+            if (fs::exists(dir) && fs::is_directory(dir)) {
+                return dir;
+            }
+        }
+
+        // Return an empty path if not found
+        CROW_LOG_WARNING << "App data directory not found! Defaulting to \"app/\""
+                         << " Searched in these paths: INSTALL_APP_DIRECTORY=" << INSTALL_APP_DIRECTORY
+                         << " LOCAL_APP_DIRECTORY=" << LOCAL_APP_DIRECTORY
+                         << " APP_DIRECTORY=" << std::getenv("APP_DIRECTORY");
+        return "app/";
+    }
```

Comment:
Shouldn't the order be reversed? First check a manual override via the environment variable, then the local build if there is one and finally the installed version?

Reply:
> Shouldn't the order be reversed? First check a manual override via the environment variable, then the local build if there is one and finally the installed version?

Generally I agree. 

The env should override for sure :+1: 

Though, I'm hesitant with `LOCAL_APP_DIRECTORY` as that is not always the local build (but the best we can get) – it is a directory defined during build time, not runtime: 
- The local build dir is written into the auto-configured `crow_config.hpp` during build  
  i.e. `LOCAL_APP_DIRECTORY=/home/blinky/arbitration_graphs/gui/app/arbitration_graphs` for a build in docker
- On runtime, the executable could be somewhere else (installed to `/usr`) on a different machine etc.

The (agreeably unrealistic, but possible) attack vector:
- arbitration_graphs GUI is run from a installed version (`/usr`)
- the machine has a `blinky` user
- running the GUI would try to expose `~/arbitration_graphs/gui/app/arbitration_graphs`, even though the user is not aware of that (you'd expect only installed app paths can be exposed)

That's why I'd prefer attempting to serve/expose a local directory, only if no installed version has been found – which will apply to devs only.
But, these would have issues, if they'd have an installed version as well. On the other hand, I'd expect them to use the dev docker :wink: 

But yeah, it's super unlikely someone with a `blinky` user uses our GUI webserver and has a local `~/arbitration_graphs/gui/app/arbitration_graphs` dir that should not be exposed :sweat_smile:

Reply:
We could experiment with alternatives to the configured header file to detect the local build, e.g.

```cmake
if (CMAKE_CURRENT_SOURCE_DIR STREQUAL CMAKE_SOURCE_DIR)
    # Define RUNNING_FROM_BUILD_DIR if Project A is the top-level project
    add_compile_definitions(RUNNING_FROM_BUILD_DIR)
endif()
```

I'm not sure though, how stable that works for our header-only repo :see_no_evil: